Add statistics and visualisations for elevation gain

waeller opened the pull request:
Hi,

Great project!

I played around a bit with the code and added some diagrams and statistics based on the elevation gain of the activities:

- on the entry page: diagram showing elevation gain on the last 30 days
- on the activity page: option to set the color of the track line based on the elevation
- on the "Summary Statistics" page: basically duplicated all the distance visualisations there for elevation gain: weekly and monthly, yearly aggregations, comulative elevation gain per year, yearly heatmap
- bubble chart page: add a chart using elevation gain
- new page "elevation eddington number" - inspired by statshunters.com: same logic than for the distance based eddington number. To make it more interesting a divisor can be selected (1, 10, 20) whicih is applied to the elevation gain. So the elevation eddington number n means that elevation n*divisor was exceeded on n days.

If you are interested in merging the PR, but prefer something done in another way, please let me know. I'll adapt it then.

waeller commented:
I merged the current refactorings from the main branch, so the branch is up-to-date again (at least for me it works ;-)

Yes, there's a lot of code duplication. I can try to minimize this in the next days.

martin-ueding commented:
Thank you for the merge conflict resolution, I know how tedious that is.

I would like to build something to generate generic plots with. Then perhaps we can remove a lot of the explicit plotting code and convert these into user-defined plots. If that works, perhaps we don't need as much code. If you want, you can wait a bit such that we don't do duplicate work.

I will not do any big refactorings now or delete code, such that we can integrate your changes when you are ready.

martin-ueding commented:
See https://github.com/martin-ueding/geo-activity-playground/issues/258 for my idea. I think with that we can already build at least some of the plots that you want to add with this PR.

waeller commented:
I've refactored the new stuff and removed all the duplications I've introduced by generalizing the existing code. I hope I did not violate to many code style conventions; I'm not really accustomed to developing with python and flask. Just tell me if I should change something.

The Plot Builder looks very interesting. But I think I need to have a look at the Vega documentation first  to understand what effect all of the parameters should have ;)

I did not check yet if we could remove code by replacing some of existing plots with Plot-Builder based ones. I'll have a look at this later.

Happy Easter Holidays :-)

martin-ueding commented on geo_activity_playground/core/activities.py:
```diff
@@ -137,21 +137,21 @@ def make_geojson_color_line(time_series: pd.DataFrame) -> str:
 
-def make_speed_color_bar(time_series: pd.DataFrame) -> dict[str, Any]:
-    low, high, clamp_speed = _make_speed_clamp(time_series["speed"])
+def make_color_bar(time_series: pd.DataFrame, column: str) -> dict[str, Any]:
```
Instead of passing `time_series` and `column`, we could consider just passing the one column data as a `pd.Series`.

waeller replied:
Ok, I'll have a look.

martin-ueding commented on geo_activity_playground/webui/blueprints/eddington_blueprints.py:
```diff
@@ -0,0 +1,266 @@
+    return blueprint
+
+
+def register_elevation_eddington_blueprint(
```
I would like to have one blueprint per file. We can make the Eddington stuff one blueprint with two routes, I think that would make sense.

martin-ueding commented on geo_activity_playground/webui/blueprints/eddington_blueprints.py:
```diff
@@ -0,0 +1,266 @@
+    search_query_history: SearchQueryHistory,
+    template_name,
+    column,
+    divisor_values_avail: list[int] = [1],
```
Why is this a list with only 1, when the from contains 10 and 20 as well?

waeller replied:
That's just a default value for the eddington number for distances where we do not want to use a divisor (or other eddingtons later?). But I can put it as an argument in the function call instead.

martin-ueding commented on geo_activity_playground/webui/columns.py:
```diff
@@ -0,0 +1,13 @@
+column_distance = {
```
These are untyped dictionaries. Perhaps a dataclass would be better?

waeller replied:
Fine with me. I'll see to it.

martin-ueding commented on geo_activity_playground/webui/templates/bubble_chart/index.html.j2:
```diff
@@ -2,16 +2,36 @@
 {% extends "page.html.j2" %}
 
 {% block container %}
+<script src="https://cdn.jsdelivr.net/npm/vega@5"></script>
```
These things should already be embedded via `page.html.j2`. If not, could you add these files to `static` such that we don't request assets from external services?

waeller replied:
Those script tags were already in the original bubble_chart index.html.j2 file, I just moved them outside the Chart div. However, I can do it.

waeller replied:
I removed the script tags, the bubble charts seem to work fine with the vega scripts included by page.html.j2, too.

martin-ueding commented:
Very nice! I've merged the changes into `main` and added a changelog entry.

waeller commented:
Thank you very much!

martin-ueding commented:
Just to let you know about a neat Python feature: You can use `@dataclasses.dataclass` to create the `__init__` method for you for such simple data classes:

https://github.com/martin-ueding/geo-activity-playground/commit/001a81c4ca169a72375143838697cde8d8cfbec3#diff-e867e7b23574ee8fd6039428d7d898f8b0e29993168c47c67f99331fae48e11dR4

waeller commented:
That's cool, didn't know that. Thanks for the hint.
